pallets/external-validators/src/lib.rs

```diff
@@ -0,0 +1,435 @@
+    Runtime: frame_system::Config + Config,
+    Fallback: tp_traits::DistributeRewards<AccountIdOf<Runtime>, CreditOf<Runtime, Currency>>,
+    Currency: frame_support::traits::fungible::Balanced<AccountIdOf<Runtime>>,
+{
+    fn distribute_rewards(
+        rewarded: AccountIdOf<Runtime>,
+        amount: CreditOf<Runtime, Currency>,
+    ) -> frame_support::pallet_prelude::DispatchResultWithPostInfo {
+        let mut total_weight = Weight::zero();
+        let collator_id = Runtime::ValidatorIdOf::convert(rewarded.clone())
+            .ok_or(Error::<Runtime>::UnableToDeriveCollatorId)?;
+        // weight to read invulnerables
+        total_weight += Runtime::DbWeight::get().reads(1);
+        if !WhitelistedValidators::<Runtime>::get().contains(&collator_id) {
+            let post_info = Fallback::distribute_rewards(rewarded, amount)?;
+            if let Some(weight) = post_info.actual_weight {
+                total_weight += weight;
+            }
+        } else {
+            Currency::resolve(&rewarded, amount).map_err(|_| TokenError::NotExpendable)?;
+            total_weight += Runtime::WeightInfo::reward_validator(
+                Runtime::MaxWhitelistedValidators::get() + Runtime::MaxExternalValidators::get(),
+            )
+        }
+        Ok(Some(total_weight).into())
+    }
+}
+
+impl<T: Config> pallet_session::SessionManager<T::ValidatorId> for Pallet<T> {
+    fn new_session(new_index: SessionIndex) -> Option<Vec<T::ValidatorId>> {
+        if new_index <= 1 {
+            return None;
+        }
+
+        let mut validators: Vec<_> = WhitelistedValidators::<T>::get().into();
+
+        if !SkipExternalValidators::<T>::get() {
+            validators.extend(ExternalValidators::<T>::get())
+        }
+
+        frame_system::Pallet::<T>::register_extra_weight_unchecked(
+            T::WeightInfo::new_session(validators.len() as u32),
+            DispatchClass::Mandatory,
+        );
+
+        Some(validators)
+    }
+
+    fn end_session(_: SessionIndex) {}
+
+    fn start_session(_start_index: SessionIndex) {}
+}
+
+impl<T: Config> pallet_session::historical::SessionManager<T::ValidatorId, ()> for Pallet<T> {
+    fn new_session(new_index: SessionIndex) -> Option<Vec<(T::ValidatorId, ())>> {
+        <Self as pallet_session::SessionManager<_>>::new_session(new_index)
+            .map(|r| r.into_iter().map(|v| (v, Default::default())).collect())
+    }
+
+    fn start_session(start_index: SessionIndex) {
+        <Self as pallet_session::SessionManager<_>>::start_session(start_index)
+    }
+
+    fn end_session(end_index: SessionIndex) {
+        <Self as pallet_session::SessionManager<_>>::end_session(end_index)
+    }
+}
+
+impl<T: Config> EraIndexProvider for Pallet<T> {
+    fn active_era() -> ActiveEraInfo {
+        <ActiveEra<T>>::get().unwrap_or(ActiveEraInfo {
+            index: 0,
+            start: None,
+        })
+    }
+}
+
+impl<T: Config> InvulnerablesProvider<T::ValidatorId> for Pallet<T> {
```

Comment:
I would call this ValidatorProvider, and return invulnerables.concat(external) if external validators are enabled, or invulnerables otherwise

Reply:
dont forget this

Reply:
ah sorry I see you have it already fixed,